### DISCUSSION TITLE
pydantic dataclass: object has no attribute '_sa_instance_state'

### DISCUSSION THREAD

**Discussion by BayerSe** [sqlalchemy/sqlalchemy#10243] [Usage Questions] [answered]
Hi,

I have seen the warning regarding alternative dataclass providers mentioned [here](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic), so if my finding below is beyond the scope of sqlalchemy, please feel free to ignore the question.

---

I'm currently working on adding a one-to-many relationship in one of my pydantic-based classes, along the lines of the following, which works perfectly with sqlalchemy v2.0.20 and pydantic v2.1.1.

```python
from __future__ import annotations
import pydantic
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Mapped, mapped_column, relationship


class Base(MappedAsDataclass, DeclarativeBase, dataclass_callable=pydantic.dataclasses.dataclass):
    ...

class Child(Base):
    __tablename__ = "child_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column(ForeignKey("parent_table.id"))


class Parent(Base):
    __tablename__ = "parent_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    children: Mapped[list[Child]] = relationship()

parent1 = Parent(id=0, children=[])
parent2 = Parent(id=1, children=[])
``` 

If I now reverse the order of the `Child` and `Parent` classes, I get an error for the *second* parent, the first one is actually instantiated successfully:

```python
from __future__ import annotations
import pydantic
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Mapped, mapped_column, relationship


class Base(MappedAsDataclass, DeclarativeBase, dataclass_callable=pydantic.dataclasses.dataclass):
    ...

class Parent(Base):
    __tablename__ = "parent_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    children: Mapped[list[Child]] = relationship()


class Child(Base):
    __tablename__ = "child_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    parent_id: Mapped[int] = mapped_column(ForeignKey("parent_table.id"))

parent1 = Parent(id=0, children=[])
parent2 = Parent(id=1, children=[])
```

This snippet raises 
```python
AttributeError: 'Parent' object has no attribute '_sa_instance_state'
```
for `parent_2` 

Thanks
Sebastian

**Comment by CaselIT** [comment]
Hi,

This seems like an issue of pydantic in how it handles forward references.

Using the stdlib dataclass raises the same issue?

**Reply by BayerSe** [reply]
Oh, sorry I forgot to mention that. stdlib's dataclass works perfectly fine.

**Comment by zzzeek** [comment]
I can't make any comment here without seeing a stack trace.   the question of *where* in pydantic this attribute is being accessed would point to what might be going wrong (but yes, it's something local to pydantic).

**Reply by BayerSe** [reply]
Sorry, @zzzeek and @CaselIT my bad. Next time I'll deliver that with the initial message. 

Here's the stacktrace

```python
Traceback (most recent call last):
  File "/home/project/scratch.py", line 22, in <module>
    parent2 = Parent(id=1, children=[])
  File "/home/project/.venv/lib/python3.10/site-packages/pydantic/_internal/_dataclasses.py", line 124, in __init__
    s.__pydantic_validator__.validate_python(ArgsKwargs(args, kwargs), self_instance=s)
  File "/home/project/.venv/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 537, in __set__
    instance_state(instance), instance_dict(instance), value, None
AttributeError: 'Parent' object has no attribute '_sa_instance_state'
```

**Reply by zzzeek** [reply, marked as answer]
so yeah, I dont know what to tell you, pydantic keeps changing what they are doing and there's no way we can keep dodging these changes.  in this case, pydantic is adding additional instrumentation to the class, **after one of the objects is instantiated**.  SQLAlchemy applies the dataclass instrumetnation to the class **first**, *then* it applies its own instrumentation.  in this case, Pydantic is coming in a **second** time and blowing that away.  they have to not do that:

```py
Base.registry.configure()

# the correct init function
print(f"Mappers are all set up: {Parent.__init__.__code__}")

parent1 = Parent(id=0, children=[])

# the **incorrect** init function
print(f"Pydantic has interfered with our mapping: {Parent.__init__.__code__}")

# so this fails
parent2 = Parent(id=1, children=[])



```

output:

```
[classic@zotac sqlalchemy:main]$ python test3.py 
Mappers are all set up: <code object __init__ at 0x7fce0446e5b0, file "<string>", line 1>
Pydantic has interfered with our mapping: <code object __init__ at 0x7fce05fb2430, file "/home/classic/.venv3/lib64/python3.11/site-packages/pydantic/_internal/_dataclasses.py", line 121>
Traceback (most recent call last):
  File "/mnt/photon_home/classic/dev/sqlalchemy/test3.py", line 32, in <module>
    parent2 = Parent(id=1, children=[])
              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/classic/.venv3/lib64/python3.11/site-packages/pydantic/_internal/_dataclasses.py", line 124, in __init__
    s.__pydantic_validator__.validate_python(ArgsKwargs(args, kwargs), self_instance=s)
  File "/mnt/photon_home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/attributes.py", line 537, in __set__
    instance_state(instance), instance_dict(instance), value, None
    ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Parent' object has no attribute '_sa_instance_state'

```

I would assume they have a step that is similar to SQLAlchemy's "configure" that is doing the cross-referencing of related classes.    However, they can't blow away `__init__` here because our `__init__` has to happen first.   They should be instrumetning `__init__` within their dataclass creation step like plain dataclasses does.

this unfortunately looks a like a bit of an impasse since one or the other project would need to re-architect their instrumetnation (and it's not going to be us, we have done the same thing for 18 years)

**Reply by BayerSe** [reply]
Thank you for looking into this, @zzzeek! This is indeed a bit unfortunate. For now, I can get away with the reordered classes but who knows what other issues might pop in future?

In the docs you write:

```
The dataclass layer of Pydantic version 1.x is not fully compatible ...
```

I guess you can change that to  ... 1.x and 2.x ... so that the above case is also covered by the warning.

Thanks
Sebastian

**Reply by CaselIT** [reply]
The funny thing is that is was compatible during the beta of the pydantic v2 version (or at least it seemed to).
A PR to change that line would be appreciated.

Regarding pydantic, they have always gone the extra mile to interfere with other libraries, so my suggestion would be to use something else if you can for classes that interact other libraries

**Reply by BayerSe** [reply]
PR: https://github.com/sqlalchemy/sqlalchemy/pull/10249

Thanks for your help! I'll reevaluate the combination of sqlalchemy and pydantic via the dataclass drop-in. Maybe there's some better way.

**Reply by CaselIT** [reply]
Thanks for the PR!

You may also take a look at the sqlmodel library, that does unify sqlalchemy and pydantic, but I'm not sure how mantained it is

**Reply by guicbrito** [reply]
Any updates on this issue? 

I've run into it using Pydantic 2.8.2 and SQLAlchemy 2.1.0b and the only solution was removing `dataclass_callable=pydantic.dataclasses.dataclass` from `Base`.

**Reply by CaselIT** [reply]
there is nothing sqlalchemy can do here, you would need to seek guidance from pydantic side

**Comment by v1a0** [comment]
### The same issue occurs with many-to-many relationships using the `secondary` feature.

### Log:

```
Media(type="photo")
Traceback (most recent call last):
  File "/.../pydevd_asyncio/pydevd_asyncio_utils.py", line 117, in _exec_async_code
    result = func()
             ^^^^^^
  File "<input>", line 1, in <module>
  File "/.../.venv/lib/python3.11/site-packages/pydantic/_internal/_dataclasses.py", line 141, in __init__
    s.__pydantic_validator__.validate_python(ArgsKwargs(args, kwargs), self_instance=s)
  File "/.../.venv/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 538, in __set__
    instance_state(instance), instance_dict(instance), value, None
    ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Media' object has no attribute '_sa_instance_state'
```



### The original code looked something like this:

```python
class Base(
    MappedAsDataclass,
    DeclarativeBase,
    dataclass_callable=pydantic.dataclasses.dataclass,
):
    pass
    

class MediaGroup(Base):
    __tablename__ = 'media_group'
    
    id: Mapped[int] = ...
    
    # Relationship
    media_items: Mapped[list["Media"]] = relationship(
        "Media",
        secondary="media_group_media",
        back_populates="groups",
    )


class Media(Base):
    __tablename__ = 'media'
    
    id: Mapped[int] = ...
    type: Mapped[str] = ...
    
    # Relationship 
    groups: Mapped[list["MediaGroup"]] = relationship(
        "MediaGroup",
        secondary="media_group_media",
        back_populates="media_items"
    )


class MediaGroupMedia(Base):
    __tablename__ = 'media_group_media'
    
    group_id: Mapped[int] = mapped_column(ForeignKey('media_group.id'), primary_key=True)
    media_id: Mapped[int] = mapped_column(ForeignKey('media.id'), primary_key=True)
    
    # Relationships 
    media_group: Mapped["MediaGroup"] = relationship("MediaGroup", back_populates="media_items")
    media: Mapped["Media"] = relationship("Media", back_populates="groups")
    
# ...

def main():
    Media(type="photo")
    
main()    
```


Requirements:

```
pydantic==2.8.2
sqlalchemy==2.0.32
```

**Reply by CaselIT** [reply]
That's an issue with how pydantic works.
I don't think it's supported at the moment since it insist in removing all extra attributes. IIRC at some point the v2 did not do that but I guess they changed that bit again

This is also the main reason why you can't map with sqlalchemy model on a pydantic class

**Reply by v1a0** [reply]
The only solution I’ve found, as mentioned above, is to stop using `pydantic.dataclasses` with SQLAlchemy models.

```python
class Base(
    MappedAsDataclass,
    DeclarativeBase,
    # dataclass_callable=pydantic.dataclasses.dataclass,
    # @v1a0: Damn f*ck pydantic... https://github.com/sqlalchemy/sqlalchemy/discussions/10243
):
    pass
```

**Reply by CaselIT** [reply]
Note that this seems to be working fine:
```py
import sqlalchemy as sa
from sqlalchemy.orm import (
    MappedAsDataclass,
    DeclarativeBase,
    mapped_column,
    relationship,
    Mapped,
)
from sqlalchemy.orm.session import Session
import pydantic


class Base(
    MappedAsDataclass,
    DeclarativeBase,
    dataclass_callable=pydantic.dataclasses.dataclass,
):
    pass


class MediaGroup(Base):
    __tablename__ = "media_group"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)

    # Relationship
    media_items: Mapped[list["Media"]] = relationship(
        "Media",
        secondary="media_group_media",
        back_populates="groups",
        default_factory=list,
    )


class Media(Base):
    __tablename__ = "media"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    type: Mapped[str]

    # Relationship
    groups: Mapped[list["MediaGroup"]] = relationship(
        "MediaGroup",
        secondary="media_group_media",
        back_populates="media_items",
        default_factory=list,
    )


class MediaGroupMedia(Base):
    __tablename__ = "media_group_media"

    group_id: Mapped[int] = mapped_column(
        sa.ForeignKey("media_group.id"), primary_key=True
    )
    media_id: Mapped[int] = mapped_column(
        sa.ForeignKey("media.id"), primary_key=True
    )

m = Media(type="photo", groups=[MediaGroup()])

e = sa.create_engine("sqlite:///")
Base.metadata.create_all(e)
with Session(e) as s:
    s.add(m)
    s.commit()

    print(m.id)

with Session(e) as s:
    x = s.get_one(Media, 1)
    print(x.groups)
print(x)
```

pydantic          2.10.3
pydantic_core     2.27.1
